aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+    return shares_to_mint;
+  }
+
+  transition deposit_public(
+    input_record: credits.aleo/credits,
+    microcredits: u64,
+    validator_address: address,
+    expected_shares: u64,
+    minimum_shares: u64
+  ) -> credits.aleo/credits {
+    // Must be a credits record because credits.aleo uses self.caller for transfers
+    let updated_record: credits.aleo/credits = credits.aleo/transfer_private_to_public(input_record, CORE_PROTOCOL, microcredits);
+    credits.aleo/bond_public(validator_address, microcredits);
+
+    return (updated_record) then finalize(self.caller, microcredits, validator_address, expected_shares, minimum_shares);
+  }
+
+  finalize deposit_public(
+    caller: address,
+    microcredits: u64,
+    validator_address: address,
+    expected_shares: u64,
+    minimum_shares: u64
+  ) {
+    let stored_validator: address = validator.get(0u8);
+    assert_eq(validator_address, stored_validator);
+
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 0u8);
+
+    // Distribute shares for new commission
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    let current_balance: u64 = total_balance.get(0u8);
+    let current_shares: u64 = total_shares.get(0u8);
+    let rewards: u64 = bonded - current_balance;
+    let commission_rate: u128 = commission_percent.get(0u8);
+    let new_commission: u64 = rewards > 0u64 ? get_commission(rewards as u128, commission_rate) : 0u64;
+    current_balance += rewards - new_commission;
+
+    let new_commission_shares: u64 = calculate_new_shares(current_balance as u128, new_commission as u128, current_shares as u128);
+    let current_commission: u64 = delegator_shares.get_or_use(ADMIN, 0u64);
+    delegator_shares.set(ADMIN, current_commission + new_commission_shares);
+
+    current_shares += new_commission_shares;
+    current_balance += new_commission;
+
+    // Calculate mint for deposit
+    let new_shares: u64 = calculate_new_shares(current_balance as u128, microcredits as u128, current_shares as u128);
+
+    // Ensure mint amount is valid
+    assert(new_shares >= 1u64);
+    assert(new_shares >= minimum_shares);
+    assert(new_shares <= expected_shares);
+
+    // Update delegator_shares mapping
+    let shares: u64 = delegator_shares.get_or_use(caller, 0u64);
+    delegator_shares.set(caller, shares + new_shares);
+
+    // Update total shares
+    total_shares.set(0u8, current_shares + new_shares);
+
+    // Update total_balance
+    total_balance.set(0u8, current_balance + microcredits);
+  }
+
+  transition withdraw_public(withdrawal_shares: u64, total_withdrawal: u64) {
+    credits.aleo/unbond_public(total_withdrawal);
+
+    return then finalize(withdrawal_shares, total_withdrawal, self.caller);
+  }
+
+  finalize withdraw_public(withdrawal_shares: u64, total_withdrawal: u64, owner: address) {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 0u8);
```

Comment:
This is dangerous -- a malicious validator could call `unbond_all` or `claim_all` and decide not to progress the protocol, effectively freezing all the funds. We should either allow withdraws to proceed or time lock `unbond_all` or `claim_all`

Reply:
Another option -- we could just not allow withdrawals during the unbond_all 360 block window. If a validator has unbonded, but has not yet re-bonded credits, we could provide a direct transition to allow users to withdraw funds directly without going through the batched withdraw process. This would circumvent a malicious validator.